docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,126 @@
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+A proposal for improving Kubelet config by versioning it and making it manageable via the API server.
+
+## Motivation
+
+The Kubelet is currently configured via command-line flags. This is painful for a number of reasons:
+- It makes it difficult to change the way Kubelets are configured in a running cluster, because it is often tedious to change the Kubelet startup configuration.
+- It makes it difficult to manage different Kubelet configurations for different nodes, e.g. if you want to canary a new config.
+- The current lack of versioned Kubelet configuration means that any time we change Kubelet flags, we risk breaking someone's setup.
+- Having the ability to pass config in the same format on-disk as via the API server will make it easier to bootstrap master nodes before the API server is available.
+
+
+## Goals of this design:
+
+- Add a well-defined, versioned configmap object for Kubelet configuration.
+- Add/Remove/Update global and per-node Kubelet config on the fly via the API server.
+- Provide a REST API to get the config currently in use by a given Kubelet.
+- We want the Kubelet to work with the built-in defaults on all distros we test against.
+- Work towards deprecating flags in favor of on-disk config and config from the API server. Master nodes will be bootstrapped from on-disk config.
+
+## Design
+
+### Definitions
+
+- `kubelet-default`: The built-in default config that every Kubelet starts up with.
+- `api-default`: The defaults provided for omitted/nil fields by the API server.
+- `on-disk-config`: `KubeletConfiguration` provided on disk, used as a fallback when the Kubelet cannot cannot contact the API server.
+- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via `on-disk-config`.
+- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
```

Comment:
@derekwaynecarr In the current example implementation, the Kubelet will try to use the Cloud Provider's implementation of `CurrentNodeName` to get the `<node-name>` portion of `<node-name>-kubelet-configuration`. If there isn't a `kcfg.Cloud` available, it will default to using the hostname (which is a `uname -n` unless a `HostnameOverride` was provided) for `<node-name>` when it looks for the configmap.